File: client/dto/accessControl.go
package dto

var AllowedPrincipalTypes = map[string]struct{}{
	"USER":            {},
	"USER_GROUP":      {},
	"SERVICE_ACCOUNT": {},
}

var AllowedScopeLevels = map[string]struct{}{
	"account":      {},
	"organization": {},
	"project":      {},
}

type AccessControlOutput[T any] struct {
	Status        string                 `json:"status,omitempty"`
	Data          T                      `json:"data,omitempty"`
	MetaData      map[string]interface{} `json:"metaData,omitempty"`
	CorrelationId string                 `json:"correlationId,omitempty"`
}

type RolesOutputData struct {
	TotalPages    int64                    `json:"totalPages,omitempty"`
	TotalItems    int64                    `json:"totalItems,omitempty"`
	PageItemCount int64                    `json:"pageItemCount,omitempty"`
	PageSize      int64                    `json:"pageSize,omitempty"`
	Content       []RolesOutputDataContent `json:"content,omitempty"`
}

type RolesOutputDataContent struct {
	Role                              RoleData `json:"role,omitempty"`
	CreatedAt                         int64    `json:"createdAt,omitempty"`
	LastModifiedAt                    int64    `json:"lastModifiedAt,omitempty"`
	RoleAssignedToUserCount           int64    `json:"roleAssignedToUserCount,omitempty"`
	RoleAssignedToUserGroupCount      int64    `json:"roleAssignedToUserGroupCount,omitempty"`
	RoleAssignedToServiceAccountCount int64    `json:"roleAssignedToServiceAccountCount,omitempty"`
}

type RoleData struct {
	Identifier  string   `json:"identifier,omitempty"`
	Name        string   `json:"name,omitempty"`
	Permissions []string `json:"permissions,omitempty"`
}

type PermissionsOutputData struct {
	Permission PermissionsOutputObject `json:"permission,omitempty"`
}

type PermissionsOutputObject struct {
	Identifier         string   `json:"identifier,omitempty"`
	Name               string   `json:"name,omitempty"`
	Status             string   `json:"status,omitempty"`
	IncludeInAllRoles  bool     `json:"includeInAllRoles,omitempty"`
	AllowedScopeLevels []string `json:"allowedScopeLevels,omitempty"`
	ResourceType       string   `json:"resourceType,omitempty"`
	Action             string   `json:"action,omitempty"`
}

type RoleAssignmentsOutputData struct {
	TotalPages    int64                              `json:"totalPages,omitempty"`
	TotalItems    int64                              `json:"totalItems,omitempty"`
	PageItemCount int64                              `json:"pageItemCount,omitempty"`
	PageSize      int64                              `json:"pageSize,omitempty"`
	Content       []RoleAssignmentsOutputDataContent `json:"content,omitempty"`
}

type RoleAssignmentsOutputDataContent struct {
	RoleAssignment RoleAssignment `json:"roleAssignment,omitempty"`
	Scope          Scope          `json:"scope,omitempty"`
	CreatedAt      int64          `json:"createdAt,omitempty"`
	LastModifiedAt int64          `json:"lastModifiedAt,omitempty"`
	HarnessManaged bool           `json:"harnessManaged,omitempty"`
}

type RoleAssignment struct {
	Identifier              string                 `json:"identifier,omitempty"`
	ResourceGroupIdentifier string                 `json:"resourceGroupIdentifier,omitempty"`
	RoleIdentifier          string                 `json:"roleIdentifier,omitempty"`
	Principal               AccessControlPrincipal `json:"principal,omitempty"`
	Managed                 bool                   `json:"managed,omitempty"`
	Internal                bool                   `json:"internal,omitempty"`
	Disabled                bool                   `json:"disabled,omitempty"`
	RoleReference           RoleReference          `json:"roleReference,omitempty"`
}

type AccessControlPrincipal struct {
	ScopeLevel string `json:"scopeLevel,omitempty"`
	Identifier string `json:"identifier,omitempty"`
	Type       string `json:"type,omitempty"`
	UniqueId   string `json:"uniqueId,omitempty"`
}

type RoleReference struct {
	Identifier string `json:"identifier,omitempty"`
	ScopeLevel string `json:"scopeLevel,omitempty"`
}

type UsersOutput struct {
	TotalPages    int64          `json:"totalPages,omitempty"`
	TotalItems    int64          `json:"totalItems,omitempty"`
	PageItemCount int64          `json:"pageItemCount,omitempty"`
	PageSize      int64          `json:"pageSize,omitempty"`
	Content       []UserInfoData `json:"content,omitempty"`
}

type UserData struct {
	Name              string `json:"name,omitempty"`
	Email             string `json:"email,omitempty"`
	UUID              string `json:"uuid,omitempty"`
	Locked            bool   `json:"locked,omitempty"`
	Disabled          bool   `json:"disabled,omitempty"`
	ExternallyManaged bool   `json:"externallyManaged,omitempty"`
}

type RoleAssignmentMetadataObject struct {
	Identifier              string `json:"identifier,omitempty"`
	RoleIdentifier          string `json:"roleIdentifier,omitempty"`
	RoleName                string `json:"roleName,omitempty"`
	RoleScopeLevel          string `json:"roleScopeLevel,omitempty"`
	ResourceGroupIdentifier string `json:"resourceGroupIdentifier,omitempty"`
	ResourceGroupName       string `json:"resourceGroupName,omitempty"`
	ManagedRole             bool   `json:"managedRole,omitempty"`
	ManagedRoleAssignment   bool   `json:"managedRoleAssignment,omitempty"`
}

type UsersRequestBody struct {
	ResourceGroupIdentifier string `json:"resourceGroupIdentifiers,omitempty"`
	RoleIdentifier          string `json:"roleIdentifiers,omitempty"`
}

type RoleInfoOutputData struct {
	Role           Role  `json:"role,omitempty"`
	Scope          Scope `json:"scope,omitempty"`
	HarnessManaged bool  `json:"harnessManaged,omitempty"`
	CreatedAt      int64 `json:"createdAt,omitempty"`
	LastModifiedAt int64 `json:"lastModifiedAt,omitempty"`
}

type Role struct {
	Identifier         string   `json:"identifier,omitempty"`
	Name               string   `json:"name,omitempty"`
	Permissions        []string `json:"permissions,omitempty"`
	AllowedScopeLevels []string `json:"allowedScopeLevels,omitempty"`
	Description        string   `json:"description,omitempty"`
}

type RoleAssignmentRequestBody struct {
	ResourceGroupFilter       []string                        `json:"resourceGroupFilter,omitempty"`
	RoleFilter                []string                        `json:"roleFilter,omitempty"`
	PrincipalTypeFilter       []string                        `json:"principalTypeFilter,omitempty"`
	PrincipalScopeLevelFilter []string                        `json:"principalScopeLevelFilter,omitempty"`
	PrincipalFilter           []RoleAssignmentPrincipalFilter `json:"principalFilter,omitempty"`
}

type UserInfoData struct {
	User                   UserData                       `json:"user,omitempty"`
	RoleAssignmentMetadata []RoleAssignmentMetadataObject `json:"roleAssignmentMetadata,omitempty"`
}

type UserGroupInfoData struct {
	AccountIdentifier string          `json:"accountIdentifier,omitempty"`
	OrgIdentifier     string          `json:"orgIdentifier,omitempty"`
	ProjectIdentifier string          `json:"projectIdentifier,omitempty"`
	Identifier        string          `json:"identifier,omitempty"`
	Name              string          `json:"name,omitempty"`
	Users             []UserBasicInfo `json:"users,omitempty"`
}

type UserBasicInfo struct {
	ID    string `json:"id,omitempty"`
	Email string `json:"email,omitempty"`
}

type ServiceAccountData struct {
	ServiceAccount               ServiceAccountInfo             `json:"serviceAccount,omitempty"`
	CreatedAt                    int64                          `json:"createdAt,omitempty"`
	LastModifiedAt               int64                          `json:"lastModifiedAt,omitempty"`
	TokensCount                  int64                          `json:"tokensCount,omitempty"`
	RoleAssignmentMetadataObject []RoleAssignmentMetadataObject `json:"roleAssignmentsMetadataDTO,omitempty"`
}

type ServiceAccountInfo struct {
	Identifier        string `json:"identifier,omitempty"`
	Name              string `json:"name,omitempty"`
	Email             string `json:"email,omitempty"`
	Description       string `json:"description,omitempty"`
	AccountIdentifier string `json:"accountIdentifier,omitempty"`
	OrgIdentifier     string `json:"orgIdentifier,omitempty"`
	ProjectIdentifier string `json:"projectIdentifier,omitempty"`
}

type RoleAssignmentPrincipalFilter struct {
	ScopeLevel string `json:"scopeLevel,omitempty"`
	Identifier string `json:"identifier,omitempty"`
	Type       string `json:"type,omitempty"`
	UniqueId   string `json:"uniqueId,omitempty"`
}

type ResourceGroup struct {
	AccountIdentifier string                      `json:"accountIdentifier,omitempty"`
	OrgIdentifier     string                      `json:"orgIdentifier,omitempty"`
	ProjectIdentifier string                      `json:"projectIdentifier,omitempty"`
	Identifier        string                      `json:"identifier,omitempty"`
	Name              string                      `json:"name,omitempty"`
	Color             string                      `json:"color,omitempty"`
	Description       string                      `json:"description,omitempty"`
	ResourceFilter    ResourceGroupResourceFilter `json:"resourceFilter,omitempty"`
}

type CreateRoleAssignmentRequestBody struct {
	Identifier              string                 `json:"identifier,omitempty"`
	ResourceGroupIdentifier string                 `json:"resourceGroupIdentifier,omitempty"`
	RoleIdentifier          string                 `json:"roleIdentifier,omitempty"`
	Principal               AccessControlPrincipal `json:"principal,omitempty"`
	Managed                 bool                   `json:"managed,omitempty"`
	Internal                bool                   `json:"internal,omitempty"`
	Disabled                bool                   `json:"disabled,omitempty"`
	RoleReference           RoleReference          `json:"roleReference,omitempty"`
}

type CreateRoleAssignmentOutputData struct {
	RoleAssignment RoleAssignment `json:"roleAssignment,omitempty"`
	Scope          Scope          `json:"scope,omitempty"`
	CreatedAt      int64          `json:"createdAt,omitempty"`
	LastModifiedAt int64          `json:"lastModifiedAt,omitempty"`
	HarnessManaged bool           `json:"harnessManaged,omitempty"`
}

type CreateResourceGroupRequestBody struct {
	ResourceGroup ResourceGroup `json:"resourceGroup,omitempty"`
}

type ResourceGroupResourceFilter struct {
	Resources []ResourceSelectorV2 `json:"resources,omitempty"`
}

type ResourceSelectorV2 struct {
	Identifiers  []string `json:"identifiers,omitempty"`
	ResourceType string   `json:"resourceType,omitempty"`
}

type CreateResourceGroupOutputData struct {
	ResourceGroup  ResourceGroup `json:"resourceGroup,omitempty"`
	CreatedAt      int64         `json:"createdAt,omitempty"`
	LastModifiedAt int64         `json:"lastModifiedAt,omitempty"`
	HarnessManaged bool          `json:"harnessManaged,omitempty"`
}

type CreateRoleOutputData struct {
	Role           Role  `json:"role,omitempty"`
	CreatedAt      int64 `json:"createdAt,omitempty"`
	LastModifiedAt int64 `json:"lastModifiedAt,omitempty"`
	HarnessManaged bool  `json:"harnessManaged,omitempty"`
	Scope          Scope `json:"scope,omitempty"`
}

type CreateUserGroupRequestBody struct {
	AccountIdentifier string   `json:"accountIdentifier,omitempty"`
	OrgIdentifier     string   `json:"orgIdentifier,omitempty"`
	ProjectIdentifier string   `json:"projectIdentifier,omitempty"`
	Identifier        string   `json:"identifier,omitempty"`
	Name              string   `json:"name,omitempty"`
	Description       string   `json:"description,omitempty"`
	Users             []string `json:"users,omitempty"`
}

type CreateUserGroupOutputData struct {
	AccountIdentifier string          `json:"accountIdentifier,omitempty"`
	OrgIdentifier     string          `json:"orgIdentifier,omitempty"`
	ProjectIdentifier string          `json:"projectIdentifier,omitempty"`
	Identifier        string          `json:"identifier,omitempty"`
	Name              string          `json:"name,omitempty"`
	Description       string          `json:"description,omitempty"`
	Users             []UserBasicInfo `json:"users,omitempty"`
}

type CreateServiceAccountRequestBody struct {
	Identifier        string `json:"identifier,omitempty"`
	Name              string `json:"name,omitempty"`
	Email             string `json:"email,omitempty"`
	Description       string `json:"description,omitempty"`
	AccountIdentifier string `json:"accountIdentifier,omitempty"`
	OrgIdentifier     string `json:"orgIdentifier,omitempty"`
	ProjectIdentifier string `json:"projectIdentifier,omitempty"`
}

type InviteUserRequestBody struct {
	Emails       []string      `json:"emails,omitempty"`
	UserGroups   []string      `json:"userGroups,omitempty"`
	RoleBindings []RoleBinding `json:"roleBindings,omitempty"`
}

type RoleBinding struct {
	ResourceGroupIdentifier string `json:"resourceGroupIdentifier,omitempty"`
	RoleIdentifier          string `json:"roleIdentifier,omitempty"`
	RoleScopeLevel          string `json:"roleScopeLevel,omitempty"`
	RoleName                string `json:"roleName,omitempty"`
	ResourceGroupName       string `json:"resourceGroupName,omitempty"`
	ManagedRole             bool   `json:"managedRole"`
}

type InviteUserOutputData struct {
	AddUserResponseMap map[string]string `json:"addUserResponseMap,omitempty"`
}

type CurrentUserData struct {
    UUID                        string      `json:"uuid,omitempty"`
    Name                        string      `json:"name,omitempty"`
    Email                       string      `json:"email,omitempty"`
    Token                       interface{} `json:"token,omitempty"`
    DefaultAccountId            string      `json:"defaultAccountId,omitempty"`
    Intent                      interface{} `json:"intent,omitempty"`
    Accounts                    []AccountInfo `json:"accounts,omitempty"`
    Admin                       bool        `json:"admin,omitempty"`
    TwoFactorAuthenticationEnabled bool     `json:"twoFactorAuthenticationEnabled,omitempty"`
    EmailVerified               bool        `json:"emailVerified,omitempty"`
    Locked                      bool        `json:"locked,omitempty"`
    Disabled                    bool        `json:"disabled,omitempty"`
    SignupAction                interface{} `json:"signupAction,omitempty"`
    Edition                     interface{} `json:"edition,omitempty"`
    BillingFrequency            interface{} `json:"billingFrequency,omitempty"`
    UTMInfo                     UTMInfo     `json:"utmInfo,omitempty"`
    ExternallyManaged           bool        `json:"externallyManaged,omitempty"`
    GivenName                   interface{} `json:"givenName,omitempty"`
    FamilyName                  interface{} `json:"familyName,omitempty"`
    ExternalId                  interface{} `json:"externalId,omitempty"`
    CreatedAt                   int64       `json:"createdAt,omitempty"`
    LastUpdatedAt               int64       `json:"lastUpdatedAt,omitempty"`
    UserPreferences             UserPreferences `json:"userPreferences,omitempty"`
    IsEnrichedInfoCollected     bool        `json:"isEnrichedInfoCollected,omitempty"`
    LastLogin                   int64       `json:"lastLogin,omitempty"`
}

type AccountInfo struct {
    UUID             string `json:"uuid,omitempty"`
    AccountName      string `json:"accountName,omitempty"`
    CompanyName      string `json:"companyName,omitempty"`
    DefaultExperience string `json:"defaultExperience,omitempty"`
    CreatedFromNG    bool   `json:"createdFromNG,omitempty"`
    NextGenEnabled   bool   `json:"nextGenEnabled,omitempty"`
}

type UTMInfo struct {
    UTMSource   interface{} `json:"utmSource,omitempty"`
    UTMContent  interface{} `json:"utmContent,omitempty"`
    UTMMedium   interface{} `json:"utmMedium,omitempty"`
    UTMTerm     interface{} `json:"utmTerm,omitempty"`
    UTMCampaign interface{} `json:"utmCampaign,omitempty"`
}

type UserPreferences struct {
    RecentSelectedScopes string      `json:"recent_selected_scopes,omitempty"`
    EnableNewNav         string      `json:"enable_new_nav,omitempty"`
    LandingPageURL       interface{} `json:"landing_page_url,omitempty"`
}

File: harness-mcp-types/src/harness/access_control.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Allowed principal types for access control
pub const ALLOWED_PRINCIPAL_TYPES: &[&str] = &["USER", "USER_GROUP", "SERVICE_ACCOUNT"];

/// Allowed scope levels for access control
pub const ALLOWED_SCOPE_LEVELS: &[&str] = &["account", "organization", "project"];

/// Generic access control output wrapper
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AccessControlOutput<T> {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<T>,
    #[serde(rename = "metaData", skip_serializing_if = "Option::is_none")]
    pub meta_data: Option<HashMap<String, serde_json::Value>>,
    #[serde(rename = "correlationId", skip_serializing_if = "Option::is_none")]
    pub correlation_id: Option<String>,
}

/// Roles output data structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RolesOutputData {
    #[serde(rename = "totalPages", skip_serializing_if = "Option::is_none")]
    pub total_pages: Option<i64>,
    #[serde(rename = "totalItems", skip_serializing_if = "Option::is_none")]
    pub total_items: Option<i64>,
    #[serde(rename = "pageItemCount", skip_serializing_if = "Option::is_none")]
    pub page_item_count: Option<i64>,
    #[serde(rename = "pageSize", skip_serializing_if = "Option::is_none")]
    pub page_size: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<Vec<RolesOutputDataContent>>,
}

/// Role output data content
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RolesOutputDataContent {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role: Option<RoleData>,
    #[serde(rename = "createdAt", skip_serializing_if = "Option::is_none")]
    pub created_at: Option<i64>,
    #[serde(rename = "lastModifiedAt", skip_serializing_if = "Option::is_none")]
    pub last_modified_at: Option<i64>,
    #[serde(rename = "roleAssignedToUserCount", skip_serializing_if = "Option::is_none")]
    pub role_assigned_to_user_count: Option<i64>,
    #[serde(rename = "roleAssignedToUserGroupCount", skip_serializing_if = "Option::is_none")]
    pub role_assigned_to_user_group_count: Option<i64>,
    #[serde(rename = "roleAssignedToServiceAccountCount", skip_serializing_if = "Option::is_none")]
    pub role_assigned_to_service_account_count: Option<i64>,
}

/// Role data structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoleData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub permissions: Option<Vec<String>>,
}

/// Permissions output data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PermissionsOutputData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub permission: Option<PermissionsOutputObject>,
}

/// Permissions output object
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PermissionsOutputObject {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub action: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resource: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub allowed: Option<bool>,
}

/// User group data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserGroupData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<HashMap<String, String>>,
    #[serde(rename = "orgIdentifier", skip_serializing_if = "Option::is_none")]
    pub org_identifier: Option<String>,
    #[serde(rename = "projectIdentifier", skip_serializing_if = "Option::is_none")]
    pub project_identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub users: Option<Vec<String>>,
    #[serde(rename = "notificationConfigs", skip_serializing_if = "Option::is_none")]
    pub notification_configs: Option<Vec<NotificationConfig>>,
    #[serde(rename = "ssoGroupId", skip_serializing_if = "Option::is_none")]
    pub sso_group_id: Option<String>,
    #[serde(rename = "ssoGroupName", skip_serializing_if = "Option::is_none")]
    pub sso_group_name: Option<String>,
    #[serde(rename = "linkedSsoId", skip_serializing_if = "Option::is_none")]
    pub linked_sso_id: Option<String>,
    #[serde(rename = "linkedSsoDisplayName", skip_serializing_if = "Option::is_none")]
    pub linked_sso_display_name: Option<String>,
    #[serde(rename = "linkedSsoType", skip_serializing_if = "Option::is_none")]
    pub linked_sso_type: Option<String>,
    #[serde(rename = "externallyManaged", skip_serializing_if = "Option::is_none")]
    pub externally_managed: Option<bool>,
}

/// Notification configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NotificationConfig {
    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
    pub notification_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub config: Option<serde_json::Value>,
}

/// Service account data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServiceAccountData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<HashMap<String, String>>,
    #[serde(rename = "orgIdentifier", skip_serializing_if = "Option::is_none")]
    pub org_identifier: Option<String>,
    #[serde(rename = "projectIdentifier", skip_serializing_if = "Option::is_none")]
    pub project_identifier: Option<String>,
}

/// Role assignment data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoleAssignmentData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub identifier: Option<String>,
    #[serde(rename = "roleIdentifier", skip_serializing_if = "Option::is_none")]
    pub role_identifier: Option<String>,
    #[serde(rename = "principalType", skip_serializing_if = "Option::is_none")]
    pub principal_type: Option<String>,
    #[serde(rename = "principalIdentifier", skip_serializing_if = "Option::is_none")]
    pub principal_identifier: Option<String>,
    #[serde(rename = "principalScopeLevel", skip_serializing_if = "Option::is_none")]
    pub principal_scope_level: Option<String>,
    #[serde(rename = "resourceGroupIdentifier", skip_serializing_if = "Option::is_none")]
    pub resource_group_identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub managed: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub disabled: Option<bool>,
}

/// Resource group data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceGroupData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<HashMap<String, String>>,
    #[serde(rename = "orgIdentifier", skip_serializing_if = "Option::is_none")]
    pub org_identifier: Option<String>,
    #[serde(rename = "projectIdentifier", skip_serializing_if = "Option::is_none")]
    pub project_identifier: Option<String>,
    #[serde(rename = "allowedScopeLevels", skip_serializing_if = "Option::is_none")]
    pub allowed_scope_levels: Option<Vec<String>>,
    #[serde(rename = "resourceFilter", skip_serializing_if = "Option::is_none")]
    pub resource_filter: Option<ResourceFilter>,
    #[serde(rename = "includedScopes", skip_serializing_if = "Option::is_none")]
    pub included_scopes: Option<Vec<ResourceScope>>,
}

/// Resource filter
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceFilter {
    #[serde(rename = "includeAllResources", skip_serializing_if = "Option::is_none")]
    pub include_all_resources: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resources: Option<Vec<ResourceFilterResource>>,
}

/// Resource filter resource
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceFilterResource {
    #[serde(rename = "resourceType", skip_serializing_if = "Option::is_none")]
    pub resource_type: Option<String>,
    #[serde(rename = "resourceIdentifier", skip_serializing_if = "Option::is_none")]
    pub resource_identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attributeFilter: Option<Vec<AttributeFilter>>,
}

/// Attribute filter
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AttributeFilter {
    #[serde(rename = "attributeName", skip_serializing_if = "Option::is_none")]
    pub attribute_name: Option<String>,
    #[serde(rename = "attributeValues", skip_serializing_if = "Option::is_none")]
    pub attribute_values: Option<Vec<String>>,
}

/// Resource scope
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceScope {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filter: Option<ScopeFilter>,
}

/// Scope filter
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScopeFilter {
    #[serde(rename = "includeChildScopes", skip_serializing_if = "Option::is_none")]
    pub include_child_scopes: Option<bool>,
}

/// User data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub uuid: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub username: Option<String>,
    #[serde(rename = "accountIdentifier", skip_serializing_if = "Option::is_none")]
    pub account_identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub locked: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub disabled: Option<bool>,
    #[serde(rename = "externallyManaged", skip_serializing_if = "Option::is_none")]
    pub externally_managed: Option<bool>,
    #[serde(rename = "twoFactorAuthenticationEnabled", skip_serializing_if = "Option::is_none")]
    pub two_factor_authentication_enabled: Option<bool>,
    #[serde(rename = "emailVerified", skip_serializing_if = "Option::is_none")]
    pub email_verified: Option<bool>,
}

/// API key data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiKeyData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<HashMap<String, String>>,
    #[serde(rename = "orgIdentifier", skip_serializing_if = "Option::is_none")]
    pub org_identifier: Option<String>,
    #[serde(rename = "projectIdentifier", skip_serializing_if = "Option::is_none")]
    pub project_identifier: Option<String>,
    #[serde(rename = "apiKeyType", skip_serializing_if = "Option::is_none")]
    pub api_key_type: Option<String>,
    #[serde(rename = "parentIdentifier", skip_serializing_if = "Option::is_none")]
    pub parent_identifier: Option<String>,
    #[serde(rename = "defaultTimeToExpireToken", skip_serializing_if = "Option::is_none")]
    pub default_time_to_expire_token: Option<i64>,
}

/// Token data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<HashMap<String, String>>,
    #[serde(rename = "apiKeyIdentifier", skip_serializing_if = "Option::is_none")]
    pub api_key_identifier: Option<String>,
    #[serde(rename = "parentIdentifier", skip_serializing_if = "Option::is_none")]
    pub parent_identifier: Option<String>,
    #[serde(rename = "apiKeyType", skip_serializing_if = "Option::is_none")]
    pub api_key_type: Option<String>,
    #[serde(rename = "validFrom", skip_serializing_if = "Option::is_none")]
    pub valid_from: Option<i64>,
    #[serde(rename = "validTo", skip_serializing_if = "Option::is_none")]
    pub valid_to: Option<i64>,
    #[serde(rename = "scheduledExpireTime", skip_serializing_if = "Option::is_none")]
    pub scheduled_expire_time: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub encoded_password: Option<String>,
}